lib/at_cmd_parser/at_cmd_parser.c

```
@@ -15,154 +15,227 @@
+			}
+
+			at_params_string_put(list, index, start_ptr, at_params_str - start_ptr);
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == STRING) {
+			char *start_ptr = at_params_str;
+
+			while (!is_quote(*at_params_str) &&
+			       !is_terminated(*at_params_str) &&
+			       !is_lfcr(*at_params_str)) {
+				at_params_str++;
+			}
+
+			at_params_string_put(list, index, start_ptr, at_params_str - start_ptr);
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == NUMBER) {
+			int value = atoi(at_params_str);
+
+			if (value <= USHRT_MAX) {
+				at_params_short_put(list, index, (u16_t)value);
+			} else {
+				at_params_int_put(list, index, value);
+			}
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == PDU) {
+			char *start_ptr = at_params_str;
+
+			while (is_valid_hex_char(*at_params_str)) {
+				at_params_str++;
+			}
+
+			at_params_string_put(list, index, start_ptr, at_params_str - start_ptr);
+
+			set_new_state(SEARCH_NEXT_PARAM);
+		} else if (state == SEARCH_NEXT_PARAM) {
+
+			while (!is_separator(*at_params_str) &&
+			       !is_terminated(*at_params_str) &&
+			       !is_lfcr(*at_params_str)) {
+				at_params_str++;
+			}
+
+			if (is_separator(*at_params_str)) {
+				at_params_str++;
+			}
+
+			index++;
+			set_new_state(DETECT_TYPE);
+		} else if (state == DETECT_TYPE) {
+			if ((index == 0) &&
+			     is_notification(*at_params_str)) {
+				set_new_state(NOTIFICATION);
+				at_params_str++;
+
+			} else if (is_number(*at_params_str)) {
+				set_new_state(NUMBER);
+
+			} else if (is_quote(*at_params_str)) {
+				at_params_str++;
+				set_new_state(STRING);
+
+			} else if (is_lfcr(*at_params_str) &&
+			          (old_state == NUMBER)) {
+
+				while (is_lfcr(*at_params_str)) {
+					at_params_str++;
+				}
+
+				set_new_state(PDU);
+			} else if ((index == 0) &&
+				    !is_notification(*at_params_str)) {
+				set_new_state(STRING);
+			} else {
+				break;
+			}
+		}
 	}
 
-	return err;
+	return 0;
```

I think you should return an error when `index != max_params`. This means the parameter array was too small or that `max_params` was too low.

---

What if call at_parser_params_from_str(), then it will default to list->param_count, but there are many response that has less than 10 params. 

However I can add an error code if I ran out of param slots, but the string is not completly parsed yet (have not hit the terminator). Which may mean that there are more parameters, but I don't have more slots in my parameter list.

---

E2BIG works for me. Thank you. Maybe we need some doc update to list the error codes?